paint/support/procedural/src/construct_runtime/parse.rs

```diff
@@ -0,0 +1,354 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use paint_support_procedural_tools::syn_ext as ext;
+use proc_macro2::Span;
+use std::collections::{BTreeMap, HashMap, HashSet};
+use syn::{
+    parse::{Parse, ParseStream},
+    spanned::Spanned,
+    token, Ident, Result, Token,
+};
+
+mod keyword {
+    syn::custom_keyword!(Block);
+    syn::custom_keyword!(NodeBlock);
+    syn::custom_keyword!(UncheckedExtrinsic);
+}
+
+#[derive(Debug)]
+pub struct RuntimeDefinition {
+    pub visibility_token: Token![pub],
+    pub enum_token: Token![enum],
+    pub name: Ident,
+    pub where_section: WhereSection,
+    pub modules: ext::Braces<ext::Punctuated<ModuleDeclaration, Token![,]>>,
+}
+
+impl Parse for RuntimeDefinition {
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(Self {
+            visibility_token: input.parse()?,
+            enum_token: input.parse()?,
+            name: input.parse()?,
+            where_section: input.parse()?,
+            modules: input.parse()?,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct WhereSection {
+    pub block: syn::TypePath,
+    pub node_block: syn::TypePath,
+    pub unchecked_extrinsic: syn::TypePath,
+}
+
+impl Parse for WhereSection {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let _: Token![where] = input.parse()?;
+        let mut seen_keys = HashSet::new();
+        let mut definitions = HashMap::new();
+        while !input.peek(token::Brace) {
```

Comment:
This `while` can be replaced with a:
```let definitions = Punctuated::parse_terminated(input)?;```

Reply:
From the `parse_terminated`: `Parsing continues until the end of this parse stream`. So this method is usually used with parsing content inside braces like `braced!` macro that effectively cuts out part of the token stream. Since we have no clear delimiters here like braces, currently I see no better way of doing it.

Reply:
Good point.